src/main/java/com/onarandombox/MultiverseCore/listeners/MVPlayerListener.java

```diff
@@ -253,112 +250,136 @@ public void playerPortalCheck(PlayerPortalEvent event) {
+     * @param world The world the player is in.
+     */
+    public void applyGameModeAndFlight(@NotNull Player player, @NotNull MVWorld world) {
+        GameModeResult keepGameModeResult = MVPlayerListener.this.actionChecker.canKeepGameMode(player, world);
+        if (keepGameModeResult.isSuccessful()) {
+            Logging.fine("Player: " + player.getName() + " can keep their gamemode. " + "Reason: " + keepGameModeResult.getName());
+            return;
+        }
+
+        // Check that the player is in the new world and they haven't been teleported elsewhere or the event cancelled.
+        if (player.getWorld() != world.getCBWorld()) {
+            Logging.fine("The gamemode/allowfly was NOT changed for player '%s' because he is now in world '%s' instead of world '%s'",
+                    player.getName(), player.getWorld().getName(), world.getName());
+            return;
+        }
+        Logging.fine("Handling gamemode for player: %s, Changing to %s", player.getName(), world.getGameMode().toString());
+        Logging.finest("From World: %s --> To World: %s", player.getWorld());
+        player.setGameMode(world.getGameMode());
+
+        // Check if their flight mode should change
+        // TODO need a override permission for this
+        if (player.getAllowFlight() && !world.getAllowFlight() && player.getGameMode() != GameMode.CREATIVE) {
+            player.setAllowFlight(false);
+            if (player.isFlying()) {
+                player.setFlying(false);
+            }
+        } else if (world.getAllowFlight()) {
+            if (player.getGameMode() == GameMode.CREATIVE) {
+                player.setAllowFlight(true);
+            }
+        }
+    }
+
+    /**
+     * Sends the reason for the teleportation failure to the sender.
+     *
+     * @param sender        The sender of the command.
+     * @param teleportee    The player being teleported.
+     * @param fromWorld     The world the player is teleporting from.
+     * @param toWorld       The world the player is teleporting to.
+     * @param result        The result of the teleportation.
+     */
+    private void tellReason(@NotNull CommandSender sender,
+                            @NotNull Player teleportee,
+                            @Nullable MVWorld fromWorld,
+                            @Nullable MVWorld toWorld,
+                            @NotNull ActionResult result
+    ) {
+        BukkitCommandIssuer issuer = this.plugin.getMVCommandManager().getCommandIssuer(sender);
+        String targetName = issuer.getIssuer() == teleportee ? "You" : teleportee.getName();
+
+        if (result.hasResult(NullPlaceResult.NULL_DESTINATION)) {
+            issuer.sendMessage(targetName + " cannot be teleported to because the destination is null.");
+        }
+        if (result.hasResult(NullPlaceResult.NULL_LOCATION)) {
+            issuer.sendMessage(targetName + " cannot be teleported to because the location is null.");
+        }
```

Comment:
I'm not crazy about all these if statements. I wonder if there's something we could do with polymorphism to clean this up a bit. Would be nice to not have all these extra lines in this class too.

Reply:
yea prob some form of mapping, but I think its fine for now. prob another thing into the todo bucket